scipy/optimize/cobyla/cobyla2.f

```diff
@@ -147,6 +147,7 @@ SUBROUTINE COBYLB (CALCFC,N,M,MPP,X,RHOBEG,RHOEND,IPRINT,MAXFUN,
          PRINT *, '  DX = ', (DX(I),I=1,N)
          PRINT *, '  BEFORE: ', N, M, (X(I),I=1,N), F, (CON(I),I=1,M)
       END IF
+      CALL CALLBACK (N,M,X)
```

Comment:
The callback should be called at the end of an iteration. I'm not familiar with Fortran, but it looks to me that the callback is being called at the start of the loop (label 40?). In general users expect that `callback` would get called the same number of times as the number of iterations. Can you check:

1. That the number of times the callback gets called is the same as the number of method iterations, `OptimizeResult.nit`?
2. That the final `x` supplied to `callback` is the same as `OptimizeResult.x`?

This might require extra tests (akin to the test you modified). If you add extra tests can you parameterize them, so they're called for the other minimize methods?

Reply:
I've written an initial fix to address this request in 454c86c. I'm having a bit of trouble setting up the build environment on my machine, so I'm relying on the CI a bit here and will address any issues that pop up.

A few notes:

1. The COBYLA algorithm doesn't set/specify `OptimizeResult.nit`. Instead it specified `OptimizeResult.nfev` (which for the COBYLA algorithm is equivalent to the number of iterations, since there is exactly one function eval per iteration). So I wrote a test to compare this value to the number of times the `callback` is called.
2. It seems there is one more place near the end of the FORTRAN subroutine which may modify `x` in some cases. However, this seems to me to be an edge case and might not be considered part of an iteration. This may cause the final value of  `x` supplied to `callback` to be different from `OptimizeResult.x` in this circumstance. I could add a call to `callback` here to fix this, but then the number of times `callback` is called may be different from `OptimizeResult.nfev`. I'm not 100% sure how to address this, since it seems to be a catch-22.

Reply:
After reading the FORTRAN code a bit better, I believe this can be solved in a way that makes sense. I'm working on this right now and will update the PR once it's done.

Reply:
This is now fixed.